### PR TITLE
Update consensus to v0.5.1

### PR DESCRIPTION
**Pull request by mkalinin**
## What was done?
This PR mainly updates consensus to [v0.5.1](https://github.com/ethereum/eth2.0-specs/tree/v0.5.1) version. 
Alongside this there are few other changes and fixes:
- Get rid of field name in `signed_root`.
- Use `signed_root` as a block id all over the code.
- Fix simulator config with empty `chainSpec` section.
- Break `SpecHelpers` down to several sub-interfaces.
- `get_shuffling2` fixed in order to pass shuffling tests.

### Spec issues fixed in this PR
- Spec has a few issues in crosslink processing part which are addressed by https://github.com/ethereum/eth2.0-specs/pull/855, https://github.com/ethereum/eth2.0-specs/pull/874.
- Overflow in `process_exit_queue`, https://github.com/ethereum/eth2.0-specs/pull/875.

### CODE REVIEW

File: core/src/main/java/org/ethereum/beacon/core/types/Gwei.java [hunk 1]
```diff
@@ -33,6 +33,23 @@ public Gwei minus(Gwei subtrahend) {
     return new Gwei(super.minus(subtrahend));
   }
 
+  public Gwei minusSat(Gwei subtrahend) {
```
Comment (zilm13) [inline core/src/main/java/org/ethereum/beacon/core/types/Gwei.java]: what does it mean?

File: start/simulator/src/main/resources/config/default-simulation-config.yml [hunk 2]
```diff
@@ -21,16 +21,16 @@ chainSpec:
...
-      LATEST_BLOCK_ROOTS_LENGTH: 64
       LATEST_RANDAO_MIXES_LENGTH: 64
       LATEST_ACTIVE_INDEX_ROOTS_LENGTH: 64
       LATEST_SLASHED_EXIT_LENGTH: 64
 
   specHelpersOptions:
-    blsVerify: false
-    blsVerifyProofOfPosession: false
-    blsSign: false
+    blsVerify: true
```
Comment (zilm13) [inline start/simulator/src/main/resources/config/default-simulation-config.yml]: why?

**Review by zilm13 — APPROVED** [review]
Minor comments

File: core/src/main/java/org/ethereum/beacon/core/types/Gwei.java [hunk 3]
```diff
@@ -33,6 +33,23 @@ public Gwei minus(Gwei subtrahend) {
...
+    if (this.compareTo(subtrahend) < 0) {
+      return Gwei.ZERO;
+    } else {
+      return minus(subtrahend);
+    }
+  }
+
+  public Gwei plusSat(Gwei addend) {
+    Gwei res = this.plus(addend);
+    if (res.lessEqual(res) && addend.greater(Gwei.ZERO)) {
```
Comment (zilm13) [inline core/src/main/java/org/ethereum/beacon/core/types/Gwei.java]: condition is always true and looks erroneous

File: consensus/src/main/java/org/ethereum/beacon/consensus/spec/EpochProcessing.java [hunk 4]
```diff
@@ -0,0 +1,1050 @@
...
+   */
+  default boolean eligible(BeaconState state, ValidatorIndex index) {
+    ValidatorRecord validator = state.getValidatorRegistry().get(index);
+    // Filter out dequeued validators
+    if (!validator.getWithdrawableEpoch().equals(getConstants().getFarFutureEpoch())) {
+      return false;
+    } else {
+      // Dequeue if the minimum amount of time has passed
+      return get_current_epoch(state).greaterEqual(
+          validator.getExitEpoch().plus(getConstants().getMinValidatorWithdrawabilityDelay()));
```
Comment (zilm13) [inline consensus/src/main/java/org/ethereum/beacon/consensus/spec/EpochProcessing.java]: we are getting overflow here.`FAR_FUTURE` which is usually exit epoch is highest possible number and we are adding `MinValidatorWithdrawabilityDelay` to it

File: core/src/main/java/org/ethereum/beacon/core/types/Gwei.java (hunk 1 above)
Comment (mkalinin) [reply core/src/main/java/org/ethereum/beacon/core/types/Gwei.java]: It's a reference to [saturation arithmetic](https://en.wikipedia.org/wiki/Saturation_arithmetic). Maybe you can come with a better naming to this method?

File: start/simulator/src/main/resources/config/default-simulation-config.yml (hunk 2 above)
Comment (mkalinin) [reply start/simulator/src/main/resources/config/default-simulation-config.yml]: oh, committed by a mistake. nice catch! 👍

File: core/src/main/java/org/ethereum/beacon/core/types/Gwei.java (hunk 3 above)
Comment (mkalinin) [reply core/src/main/java/org/ethereum/beacon/core/types/Gwei.java]: another good catch!

File: chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java [hunk 5]
```diff
@@ -261,22 +253,21 @@ private void updateCurrentObservableState(BeaconTupleDetails head, SlotNumber sl
...
-   * Applies next slot transitions until the <code>targetSlot</code> but
-   * doesn't apply EpochTransition for the <code>targetSlot</code>
+   * Applies next slot transitions until the <code>targetSlot</code>.
    *
    * @param source Source state
+   * @param targetSlot target slot.
    * @return new state, result of applied transition to the latest input state
    */
-  private BeaconStateEx applySlotTransitionsWithoutEpoch(
+  private BeaconStateEx applySlotTransitions(
```
[the remark this answers is not in the document]
Comment (mkalinin) [reply chain/src/main/java/org/ethereum/beacon/chain/observer/ObservableStateProcessorImpl.java]: There was an idea to add a transition to handle this particular case. Will try to hash it out.

File: core/src/main/java/org/ethereum/beacon/core/types/Gwei.java (hunk 1 above)
Comment (zilm13) [reply core/src/main/java/org/ethereum/beacon/core/types/Gwei.java]: Just add javadoc comment to `plusSat` and `minSat` like in `UInt64`

File: consensus/src/main/java/org/ethereum/beacon/consensus/spec/EpochProcessing.java (hunk 4 above)
Comment (mkalinin) [reply consensus/src/main/java/org/ethereum/beacon/consensus/spec/EpochProcessing.java]: 👍

**Comment by mkalinin** [comment]
@zilm13 ready for the next round

**Review by zilm13 — APPROVED** [review]
LGTM

File: consensus/src/main/java/org/ethereum/beacon/consensus/util/CachingBeaconChainSpec.java [hunk 6]
```diff
@@ -2,57 +2,48 @@
...
-public class CachingSpecHelpers extends SpecHelpers {
-
-  public static SpecHelpers createWithSSZHasher(@Nonnull SpecConstants constants) {
-    Objects.requireNonNull(constants);
-
-    Function<BytesValue, Hash32> hashFunction = Hashes::keccak256;
-    ObjectHasher<Hash32> sszHasher = SSZObjectHasher.create(hashFunction);
-    return new SpecHelpers(constants, hashFunction, sszHasher);
-  }
+public class CachingBeaconChainSpec extends DelegatingBeaconChainSpec {
```
Comment (Nashatyrev) [inline consensus/src/main/java/org/ethereum/beacon/consensus/util/CachingBeaconChainSpec.java]: Delegating wouldn't work for polymorphic invocations. E.g. `CachingSpecHelpers.get_permuted_list` would never be called

**Review by Nashatyrev — CHANGES_REQUESTED** [review]

File: consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java [hunk 7]
```diff
@@ -2105,6 +2105,9 @@ private Gwei get_inactivity_penalty(BeaconState state, ValidatorIndex index, Epo
         new Gwei[state.getValidatorRegistry().size().getIntValue()],
         new Gwei[state.getValidatorRegistry().size().getIntValue()]
     };
+    for (ValidatorIndex index : state.getValidatorRegistry().size()) {
+      deltas[0][index.getIntValue()] = deltas[1][index.getIntValue()] = Gwei.ZERO;
+    }
```
[the remark this answers is not in the document]
Comment (mkalinin) [reply consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java]: Definitely 👍

File: consensus/src/main/java/org/ethereum/beacon/consensus/util/CachingBeaconChainSpec.java (hunk 6 above)
Comment (mkalinin) [reply consensus/src/main/java/org/ethereum/beacon/consensus/util/CachingBeaconChainSpec.java]: That's true. Once call reaches delegate it never goes back to a class originated this call. Let me think around a little bit.

**Review by Nashatyrev — APPROVED** [review]

File: consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java [hunk 8]
```diff
@@ -2521,14 +2521,15 @@ public void update_registry_and_shuffling_data(MutableBeaconState state) {
...
+        ) % SHARD_COUNT
         state.current_shuffling_seed = generate_seed(state, state.current_shuffling_epoch) */
       update_validator_registry(state);
       // If we update the registry, update the shuffling data and shards as well
       state.setCurrentShufflingEpoch(next_epoch);
-      state.setCurrentShufflingStartShard(state.getCurrentShufflingStartShard().plusModulo(
-          get_current_epoch_committee_count(state), constants.getShardCount()));
+      state.setCurrentShufflingStartShard(ShardNumber.of(
+          state.getCurrentShufflingStartShard().plus(get_current_epoch_committee_count(state))
+              .modulo(constants.getShardCount())));
```
[the remark this answers is not in the document]
Comment (Nashatyrev) [reply consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java]: BTW still wondering...

[the remark this answers is not in the document]
Comment (mkalinin) [reply consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java]: hm, I brought it back: https://github.com/harmony-dev/beacon-chain-java/pull/113/files#diff-393fe29c319fa1ad2601d34ab8c1e311R870

[the remark this answers is not in the document]
Comment (Nashatyrev) [reply consensus/src/main/java/org/ethereum/beacon/consensus/SpecHelpers.java]: 👍

**Pull request merged**